datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,153 @@
+   *                           instances, before triggering a rebalance. The default is
+   *                           {@value DEFAULT_IMBALANCE_THRESHOLD}.
+   */
+  public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<Integer> imbalanceThreshold) {
+    super(maxTasks, imbalanceThreshold);
+  }
+  /**
+   * assign partitions to a particular datastream group
+   *
+   * @param currentAssignment the old assignment
+   * @param datastreamPartitions the subscribed partitions for the particular datastream group
+   * @return new assignment mapping
+   */
+  public Map<String, Set<DatastreamTask>> assignPartitions(Map<String,
+      Set<DatastreamTask>> currentAssignment, DatastreamGroupPartitionsMetadata datastreamPartitions) {
+
+    LOG.debug("old partition assignment info, assignment: {}", currentAssignment);
+
+    String dgName = datastreamPartitions.getDatastreamGroup().getName();
+
+    // Step 1: collect the # of tasks and figured out the unassigned partitions
+    List<String> assignedPartitions = new ArrayList<>();
+    int totalTaskCount = 0;
+    for (Set<DatastreamTask> tasks : currentAssignment.values()) {
+      Set<DatastreamTask> dgTask = tasks.stream().filter(t -> dgName.equals(t.getTaskPrefix())).collect(Collectors.toSet());
+      dgTask.stream().forEach(t -> assignedPartitions.addAll(t.getPartitionsV2()));
+      totalTaskCount += dgTask.size();
+    }
+
+    List<String> unassignedPartitions = new ArrayList<>(datastreamPartitions.getPartitions());
+    unassignedPartitions.removeAll(assignedPartitions);
+
+    int maxPartitionPerTask = datastreamPartitions.getPartitions().size() / totalTaskCount;
+    // calculate how many tasks are allowed to have slightly more partitions
+    final AtomicInteger remainder = new AtomicInteger(datastreamPartitions.getPartitions().size() % totalTaskCount);
+    LOG.debug("maxPartitionPerTask {}, task count {}", maxPartitionPerTask, totalTaskCount);
+
+    Collections.shuffle(unassignedPartitions);
+
+    Map<String, Set<DatastreamTask>> newAssignment = new HashMap<>();
+
+    //Step 2: generate new assignment. Assign unassigned partitions to tasks and create new task if there is
+    // a partition change
+    currentAssignment.keySet().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+      Set<DatastreamTask> newAssignedTask = tasks.stream().map(task -> {
+        if (!dgName.equals(task.getTaskPrefix())) {
+          return task;
+        } else {
+          Set<String> newPartitions = new HashSet<>(task.getPartitionsV2());
+          newPartitions.retainAll(datastreamPartitions.getPartitions());
+
+          //We need to create new task if the partition is changed
+          boolean partitionChanged = newPartitions.size() != task.getPartitionsV2().size();
+
+          int allowedPartitions = remainder.get() > 0 ? maxPartitionPerTask + 1 : maxPartitionPerTask;
```

Comment:
As mentioned earlier, ceil will avoid this code.

Reply:
same above, this is probably necessary for my current algorithm